lib/sharded-adapter.ts

```diff
@@ -128,7 +128,8 @@ class ShardedRedisAdapter extends ClusterAdapter {
       this.opts.subscriptionMode === "dynamic" &&
       message.type === MessageType.BROADCAST &&
       message.data.requestId === undefined &&
-      message.data.opts.rooms.length === 1;
+      message.data.opts.rooms.length === 1 &&
+      !message.data.opts.flags.expectSingleResponse;
```

Comment:
I'm not sure this is sufficient, since this will only work with RemoteSocket objects, and not when calling `io.to(someSocketId).emit(...)`.

Something like:

```js
const useDynamicChannel =
      this.opts.subscriptionMode === "dynamic" &&
      message.type === MessageType.BROADCAST &&
      message.data.requestId === undefined &&
      message.data.opts.rooms.length === 1 &&
      this.rooms.has(message.data.opts.rooms[0]) &&
      !this.sids.has(message.data.opts.rooms[0]);
```

should work to detect public rooms, but it might exclude public rooms that do not exist on the `io` instance, which kind of defeats the purpose of the dynamic mode...

Reply:
Yes, that's right, I forgot to mention this doesn't fix `.to(id).emit()`, only `socket.emit()` on remote socket instances. I also wasn't sure how big of an issue that part was because the [documentation very clearly says](https://socket.io/docs/v4/server-api/#socketid) you should not use the ID directly in your application - but I have to admit we did exactly that in some places of our app.

I would definitely prefer to fix that case as well, and `this.sids.has()` was the first thing I tried, but it seemed to me that `sids` only contains the local IDs, not remote ones, so it didn't actually fix anything.

Given the information available in the adapter instance, I couldn't think of anything better than checking for the ID format, e.g. "if it has exactly this many characters, it's a private ID" but that would obviously have some false positives.